internal/collector/processor/idbatchqueue/id_batch_queue.go

```diff
@@ -0,0 +1,149 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package idbatchqueue defines a FIFO queue of fixed size in which the
+// elements are batches of ids defined as byte slices.
+package idbatchqueue
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	// ErrInvalidNumBatches occurs when an invalid number of batches is specified.
+	ErrInvalidNumBatches = errors.New("invalid number of batches, it must be greater than zero")
+	// ErrInvalidBatchChannelSize occurs when an invalid batch channel size is specified.
+	ErrInvalidBatchChannelSize = errors.New("invalid batch channel size, it must be greater than zero")
+)
+
+// ID is the type of each element in the batch.
+type ID []byte
+
+// Batch is the type of batches held by the Batcher.
+type Batch []ID
+
+// Batcher behaves like a FIFO queue of a fixed number of batches, in which
+// items can be enqueued concurrently. The caller is in control of when a batch is
+// completed and a new one should be started.
+type Batcher interface {
+	// Enqueue puts the given id on the batch being currently built. The client is in charge
+	// of limiting the growth of the current batch if appropriate for its scenario. It can
+	// either call Dequeue earlier or stop adding new items depending on what is required by
+	// the scenario.
+	Enqueue(id ID)
+	// Dequeue takes the batch at the front of the queue, and moves the current batch to
+	// the end of the queue, creating a new batch to receive new enqueued items.
+	// It returns the batch that was in front of the queue and a boolean
+	// that if true indicates that there are more batches to be retrieved.
+	Dequeue() (Batch, bool)
+	// Stop informs that no more items are going to be enqueued. After
+	// this method is called attempts to enqueue new items will panic.
+	Stop()
+}
+
+var _ Batcher = (*batcher)(nil)
+
+type batcher struct {
+	pendingIds chan ID    // Channel for the ids to be added to the next batch.
+	batches    chan Batch // Channel with already captured batches.
+
+	// cbMutex protects the currentBatch storing ids.
+	cbMutex      sync.Mutex
+	currentBatch Batch
+
+	numBatches                uint64
+	newBatchesInitialCapacity uint64
+	stopchan                  chan bool
+	stopped                   bool
+}
+
+// New creates a Batcher that will hold numBatches in its FIFO queue, having a channel with
+// batchChannelSize to receive new items. New batches will be created with capacity set to
+// newBatchesInitialCapacity.
+func New(numBatches, newBatchesInitialCapacity, batchChannelSize uint64) (Batcher, error) {
+	if numBatches < 1 {
+		return nil, ErrInvalidNumBatches
+	}
+	if batchChannelSize < 1 {
```

Comment:
why not similar check for newBatchesInitialCapacity?

Reply:
newBatchesInitialCapacity is to allow some guess about the typical size of batches in order to reduce re-allocations when the batches are growing and the code works even if it is zero. The other two can cause deadlock if their value is zero.